src/System.Binary.Base64/System/Binary/Base64Decoder.cs

```diff
@@ -0,0 +1,266 @@
+            else
+            {
+                return baseLength;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int Decode(ref byte srcBytes, ref sbyte decodingMap)
+        {
+            int i0 = srcBytes;
+            int i1 = Unsafe.Add(ref srcBytes, 1);
+            int i2 = Unsafe.Add(ref srcBytes, 2);
+            int i3 = Unsafe.Add(ref srcBytes, 3);
+
+            i0 = Unsafe.Add(ref decodingMap, i0);
+            i1 = Unsafe.Add(ref decodingMap, i1);
+            i2 = Unsafe.Add(ref decodingMap, i2);
+            i3 = Unsafe.Add(ref decodingMap, i3);
+
+            i0 <<= 18;
+            i1 <<= 12;
+            i2 <<= 6;
+
+            i0 |= i3;
+            i1 |= i2;
+
+            i0 |= i1;
+            return i0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int DecodePadByOne(ref byte srcBytes, ref sbyte decodingMap)
+        {
+            int i0 = srcBytes;
+            int i1 = Unsafe.Add(ref srcBytes, 1);
+            int i2 = Unsafe.Add(ref srcBytes, 2);
+
+            i0 = Unsafe.Add(ref decodingMap, i0);
+            i1 = Unsafe.Add(ref decodingMap, i1);
+            i2 = Unsafe.Add(ref decodingMap, i2);
+
+            i0 <<= 18;
+            i1 <<= 12;
+            i2 <<= 6;
+            
+            i1 |= i2;
+
+            i0 |= i1;
+            return i0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int DecodePadByTwo(ref byte srcBytes, ref sbyte decodingMap)
+        {
+            int i0 = srcBytes;
+            int i1 = Unsafe.Add(ref srcBytes, 1);
+
+            i0 = Unsafe.Add(ref decodingMap, i0);
+            i1 = Unsafe.Add(ref decodingMap, i1);
+
+            i0 <<= 18;
+            i1 <<= 12;
+
+            i0 |= i1;
+            return i0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void WriteThreeBytes(ref byte destBytes, int i0)
+        {
+            destBytes = (byte)(i0 >> 16);
+            Unsafe.Add(ref destBytes, 1) = (byte)(i0 >> 8);
+            Unsafe.Add(ref destBytes, 2) = (byte)i0;
+        }
+
+        public TransformationStatus Transform(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesConsumed, out int bytesWritten)
+        {
+            ref byte srcBytes = ref source.DangerousGetPinnableReference();
+            ref byte destBytes = ref destination.DangerousGetPinnableReference();
+
+            int srcLength = source.Length / 4 * 4;  // only decode input up to the closest multiple of 4.
+            int destLength = destination.Length;
+
+            int sourceIndex = 0;
+            int destIndex = 0;
+
+            if (source.Length == 0) goto DoneExit;
+
+            ref sbyte decodingMap = ref s_decodingMap[0];
+
+            while (sourceIndex < srcLength - 4) // TODO: you can make this more efficient by using unsafe pointer as the loop limit
```

Comment:
> you can make this more efficient by using unsafe pointer as the loop limit

Can you show me a code sample on how to do that?

> you can get rid off this line by including it in loop limit

If I move this check into the loop limit, won't I have to add an extra if check after the loop to see if we need to go to DestinationSmallExit?
`if (destIndex > destLength - 3) goto DestinationSmallExit;`